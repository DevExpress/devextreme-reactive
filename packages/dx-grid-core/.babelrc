{
  "presets": [
    [ "env", { "targets": { "browsers": ["last 2 versions"] } } ]
  ],
  "plugins": [
    "transform-object-rest-spread"
  ]
}
