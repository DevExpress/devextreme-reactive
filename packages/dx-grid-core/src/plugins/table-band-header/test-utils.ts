import { TABLE_DATA_TYPE } from '../table/constants';

/** @internal */
export const generateColumns = (length: number) => (
  Array.from({ length }).map((_, index) => ({
    key: `${index}`,
    column: { name: `${index}` },
    type: TABLE_DATA_TYPE,
  }))
);

/** @internal */
export const generateChains = (
  columnCount: number, compressed: any[],
) => compressed.map((chainBands) => {
  const columns = generateColumns(columnCount);
  const startTitleMap = {};
  const startPoints = new Set([0]);
  const endPoints = new Set([columns.length - 1]);

  Object.keys(chainBands).forEach((title) => {
    const [start, end] = chainBands[title];
    startPoints.add(start);
    endPoints.add(end);
    if (0 < start) {
      endPoints.add(start - 1);
    }
    if (end < columns.length - 1) {
      startPoints.add(end + 1);
    }
    startTitleMap[start] = title;
  });
  const indexes = Array.from(startPoints).concat(Array.from(endPoints))
    .sort((a, b) => a - b)
    .filter(p => p < columns.length);

  const chain = [];
  for (let i = 0; i < indexes.length - 1; i += 2) {
    const start = indexes[i];
    const end = indexes[i + 1];
    chain.push({
      bandTitle: startTitleMap[start] || null,
      columns: columns.slice(start, end + 1),
      start,
    });
  }

  return chain;
});
