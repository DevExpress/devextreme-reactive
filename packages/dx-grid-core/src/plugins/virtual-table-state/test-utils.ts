import { Interval } from '../../types';

export const createInterval = (start: number, end: number) => ({ start, end });
export const generateRows = (interval: Interval, type = 'rows') => (
  Array
    .from({ length: interval.end - interval.start })
    .map((_, i) => ({ id: interval.start + i, type }))
);
export const createVirtualRows = (interval: Interval) => ({
  start: interval.start,
  rows: generateRows(interval),
});
