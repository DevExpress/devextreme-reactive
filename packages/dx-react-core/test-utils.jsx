import * as React from 'react';
import {
  Getter,
  Action,
  Template,
  TemplatePlaceholder,
  TemplateConnector,
  Plugin,
} from './';

const computedEntries = object => Object.getOwnPropertyNames(object)
  .reduce((acc, key) => Object.assign(acc, { [key]: object[key] }), {});

let actionExecutor = () => {};
// eslint-disable-next-line react/prop-types
const ComputedStateContainer = ({ actions }) => (
  <button
    className="actionExecutor"
    onClick={() => actionExecutor(actions)}
  />
);

export const pluginDepsToComponents = (
  deps,
  depsOverrides = {},
) => (
  <Plugin>
    {[...(deps.plugins || []), ...(depsOverrides.plugins || [])]
      .map(plugin => <Plugin name={plugin} key={plugin}><div /></Plugin>)}
    {Object.entries({ ...deps.getter, ...depsOverrides.getter })
      .map(([name, value]) => <Getter key={`getter_${name}`} name={name} value={value} />)}
    {Object.entries({ ...deps.action, ...depsOverrides.action })
      .map(([name, action]) => <Action key={`action_${name}`} name={name} action={action} />)}
    {Object.entries({ ...deps.template, ...depsOverrides.template })
      .map(([name, templateParams]) => (
        <Template key={`template_${name}`} name="root">
          <div>
            <TemplatePlaceholder name={name} params={templateParams} />
            <TemplatePlaceholder />
          </div>
        </Template>
      ))}
    <Template name="root">
      <TemplateConnector>
        {(getters, actions) => (
          <ComputedStateContainer
            getters={computedEntries(getters)}
            actions={computedEntries(actions)}
          />
        )}
      </TemplateConnector>
      <TemplatePlaceholder />
    </Template>
  </Plugin>
);

export const getComputedState = tree => tree.find(ComputedStateContainer).props().getters;

export const executeComputedAction = (tree, executor) => {
  actionExecutor = executor;
  tree.find(ComputedStateContainer).find('.actionExecutor').simulate('click');
};
