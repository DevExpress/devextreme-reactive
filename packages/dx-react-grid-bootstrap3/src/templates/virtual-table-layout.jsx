import React from 'react';
import PropTypes from 'prop-types';
import {
  TableLayout,
  VirtualTableLayout as VirtualTableLayoutCore,
} from '@devexpress/dx-react-grid';
import { TableContainer } from './table-container';
import { Table } from './table';

const MINIMAL_COLUMN_WIDTH = 120;

const HeadTable = props => <Table use="head" {...props} />;
const TableHead = props => <thead {...props} />;
const TableBody = props => <tbody {...props} />;

export const VirtualTableLayout = ({
  headerRows,
  bodyRows,
  columns,
  cellComponent,
  rowComponent,
  height,
  estimatedRowHeight,
}) => (
  <TableLayout
    layoutComponent={VirtualTableLayoutCore}
    headerRows={headerRows}
    rows={bodyRows}
    columns={columns}
    cellComponent={cellComponent}
    rowComponent={rowComponent}
    headComponent={TableHead}
    bodyComponent={TableBody}
    tableComponent={Table}
    headTableComponent={HeadTable}
    containerComponent={TableContainer}
    estimatedRowHeight={estimatedRowHeight}
    minColumnWidth={MINIMAL_COLUMN_WIDTH}
    height={height}
  />
);

VirtualTableLayout.propTypes = {
  headerRows: PropTypes.array.isRequired,
  bodyRows: PropTypes.array.isRequired,
  columns: PropTypes.array.isRequired,
  cellComponent: PropTypes.func.isRequired,
  rowComponent: PropTypes.func.isRequired,
  height: PropTypes.number.isRequired,
  estimatedRowHeight: PropTypes.number.isRequired,
};
