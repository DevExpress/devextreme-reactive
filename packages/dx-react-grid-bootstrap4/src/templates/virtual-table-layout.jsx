import * as React from 'react';
import {
  TableLayout,
  VirtualTableLayout as VirtualTableLayoutCore,
} from '@devexpress/dx-react-grid';

const MINIMAL_COLUMN_WIDTH = 150;

export const VirtualTableLayout = props => (
  <TableLayout
    layoutComponent={VirtualTableLayoutCore}
    minColumnWidth={MINIMAL_COLUMN_WIDTH}
    {...props}
  />
);
