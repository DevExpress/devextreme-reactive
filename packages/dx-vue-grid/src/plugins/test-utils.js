import {
  DxGetter,
  DxAction,
  DxTemplate,
  DxTemplatePlaceholder,
  DxTemplateConnector,
  DxPlugin,
} from '@devexpress/dx-vue-core';

const computedEntries = object => Object.getOwnPropertyNames(object)
  .reduce((acc, key) => Object.assign(acc, { [key]: object[key] }), {});

let actionExecutor = () => {};
// eslint-disable-next-line react/prop-types
const ComputedStateContainer = {
  name: 'ComputedStateContainer',
  props: {
    getters: {
      type: Object,
    },
    actions: {
      type: Object,
    },
  },
  render() {
    const { actions } = this;
    return (
      <button
        class="actionExecutor"
        onClick={() => actionExecutor(actions)}
      />
    );
  },
};

export const PluginDepsToComponents = {
  props: {
    deps: {
      type: Object,
      default: () => ({}),
    },
    depsOverrides: {
      type: Object,
      default: () => ({}),
    },
  },
  render() {
    const {
      deps,
      depsOverrides,
    } = this;

    return (
      <DxPlugin>
        {[...(deps.plugins || []), ...(depsOverrides.plugins || [])]
          .map(plugin => <DxPlugin name={plugin} key={plugin}><div /></DxPlugin>)}
        {Object.entries({ ...deps.getter, ...depsOverrides.getter })
          .map(([name, value]) => <DxGetter key={`getter_${name}`} name={name} value={value} />)}
        {Object.entries({ ...deps.action, ...depsOverrides.action })
          .map(([name, action]) => <DxAction key={`action_${name}`} name={name} action={action} />)}
        {Object.entries({ ...deps.template, ...depsOverrides.template })
          .map(([name, templateParams]) => (
            <DxTemplate key={`template_${name}`} name="root">
              <div>
                <DxTemplatePlaceholder name={name} params={templateParams} />
                <DxTemplatePlaceholder />
              </div>
            </DxTemplate>
          ))}
        <DxTemplate name="root">
          <div>
            <DxTemplateConnector>
              {({ getters, actions }) => (
                <ComputedStateContainer
                  getters={computedEntries(getters)}
                  actions={computedEntries(actions)}
                />
              )}
            </DxTemplateConnector>
            <DxTemplatePlaceholder />
          </div>
        </DxTemplate>
      </DxPlugin>
    );
  },
};

export const getComputedState = tree => tree.find(ComputedStateContainer).props().getters;

export const executeComputedAction = (tree, executor) => {
  actionExecutor = executor;
  tree.find(ComputedStateContainer).find('.actionExecutor').trigger('click');
};
