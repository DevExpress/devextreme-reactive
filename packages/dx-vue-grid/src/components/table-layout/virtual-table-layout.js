import { getCollapsedGrid } from '@devexpress/dx-grid-core';
import { DxSizer, DxRefHolder } from '@devexpress/dx-vue-core';
import { ColumnGroup } from './column-group';

export const VirtualTableLayout = {
  name: 'VirtualTableLayout',
  props: {
    minWidth: {
      type: Number,
      required: true,
    },
    minColumnWidth: {
      type: Number,
      required: true,
    },
    height: {
      type: Number,
      required: true,
    },
    headerRows: {
      type: Array,
      default: () => [],
    },
    bodyRows: {
      type: Array,
      required: true,
    },
    columns: {
      type: Array,
      required: true,
    },
    cellComponent: {
      type: Object,
      required: true,
    },
    rowComponent: {
      type: Object,
      required: true,
    },
    bodyComponent: {
      type: Object,
      required: true,
    },
    headComponent: {
      type: Object,
      default: () => null,
    },
    tableComponent: {
      type: Object,
      required: true,
    },
    headTableComponent: {
      type: Object,
      default: () => null,
    },
    containerComponent: {
      type: Object,
      required: true,
    },
    estimatedRowHeight: {
      type: Number,
      required: true,
    },
    getCellColSpan: {
      type: Function,
      required: true,
    },
  },
  data() {
    return ({
      rowHeights: new Map(),
      viewportTop: 0,
      viewportLeft: 0,
    });
  },
  mounted() {
    this.storeRowHeights();
  },
  updated() {
    this.storeRowHeights();
  },
  watch: {
    headerRows(headerRows) {
      const { rowHeights: prevRowHeight, bodyRows } = this;
      const rowHeights = [headerRows, bodyRows].reduce(
        (acc, row) => {
          const rowHeight = prevRowHeight.get(row.key);
          if (rowHeight !== undefined) {
            acc.set(row.key, rowHeight);
          }
          return acc;
        },
        new Map(),
      );
      this.rowHeights = rowHeights;
    },
    bodyRows(bodyRows) {
      const { rowHeights: prevRowHeight, headerRows } = this;
      const rowHeights = [headerRows, bodyRows].reduce(
        (acc, row) => {
          const rowHeight = prevRowHeight.get(row.key);
          if (rowHeight !== undefined) {
            acc.set(row.key, rowHeight);
          }
          return acc;
        },
        new Map(),
      );
      this.rowHeights = rowHeights;
    },
  },
  methods: {
    getRowHeight(row) {
      const storedHeight = this.rowHeights.get(row.key);
      if (storedHeight !== undefined) return storedHeight;
      if (row.height) return row.height;
      return this.estimatedRowHeight;
    },
    storeRowHeights() {
      const rowsWithChangedHeights = Object.keys(this.$refs)
        .map(node => this.$refs[node])
        .filter(node => !!node)
        .map(node => [node.$attrs.row, node.$el.getBoundingClientRect().height])
        .filter(([row, height]) => height !== this.getRowHeight(row));

      if (rowsWithChangedHeights.length) {
        const newRowHeights = new Map();
        rowsWithChangedHeights
          .forEach(([row, height]) => newRowHeights.set(row.key, height));

        this.rowHeights = newRowHeights;
      }
    },
    updateViewport(e) {
      const node = e.target;

      if (node !== e.currentTarget) {
        return;
      }

      // NOTE: prevent iOS to flicker in bounces
      if (node.scrollTop < 0 ||
        node.scrollLeft < 0 ||
        node.scrollLeft + node.clientWidth > Math.max(node.scrollWidth, node.clientWidth) ||
        node.scrollTop + node.clientHeight > Math.max(node.scrollHeight, node.clientHeight)) {
        return;
      }

      this.viewportTop = node.scrollTop;
      this.viewportLeft = node.scrollLeft;
    },
    renderRowsBlock(collapsedGrid, Table, Body) {
      const {
        minWidth,
        rowComponent: Row,
        cellComponent: Cell,
      } = this;

      return (
        <Table
          style={{ minWidth: `${minWidth}px` }}
          minWidth={`${minWidth}px`}
        >
          <ColumnGroup
            columns={collapsedGrid.columns}
          />
          <Body>
            {collapsedGrid.rows.map((visibleRow) => {
              const { row, cells = [] } = visibleRow;
              return (
                <DxRefHolder
                  key={row.key}
                  ref={`${row.key}`}
                  row={row}
                >
                  <Row
                    tableRow={row}
                    height={row.height !== undefined ? `${row.height}px` : undefined}
                  >
                    {cells.map((cell) => {
                      const { column } = cell;
                      return (
                        <Cell
                          key={column.key}
                          tableRow={row}
                          tableColumn={column}
                          style={column.animationState}
                          colSpan={cell.colSpan}
                        />
                      );
                    })}
                  </Row>
                </DxRefHolder>
              );
            })}
          </Body>
        </Table>
      );
    },
  },
  render() {
    const {
      headerRows,
      bodyRows,
      columns,
      minColumnWidth,
      height,
      containerComponent: Container,
      headTableComponent: HeadTable,
      tableComponent: Table,
      headComponent: Head,
      bodyComponent: Body,
      getCellColSpan,
      updateViewport,
      renderRowsBlock,
    } = this;

    return (
      <DxSizer>
        {({ width }) => {
          const headHeight = headerRows.reduce((acc, row) => acc + this.getRowHeight(row), 0);
          const getColSpan = (tableRow, tableColumn) =>
            getCellColSpan({ tableRow, tableColumn, tableColumns: columns });
          const collapsedHeaderGrid = getCollapsedGrid({
            rows: headerRows,
            columns,
            top: 0,
            left: this.viewportLeft,
            width,
            height: headHeight,
            getColumnWidth: column => column.width || minColumnWidth,
            getRowHeight: this.getRowHeight,
            getColSpan,
          });
          const collapsedBodyGrid = getCollapsedGrid({
            rows: bodyRows,
            columns,
            top: this.viewportTop,
            left: this.viewportLeft,
            width,
            height: height - headHeight,
            getColumnWidth: column => column.width || minColumnWidth,
            getRowHeight: this.getRowHeight,
            getColSpan,
          });
          return (
            <Container
              style={{ height: `${height}px` }}
              onScroll={updateViewport}
            >
              {!!headerRows.length && renderRowsBlock(collapsedHeaderGrid, HeadTable, Head)}
              {renderRowsBlock(collapsedBodyGrid, Table, Body)}
            </Container>
          );
        }}
      </DxSizer>
    );
  },
};
