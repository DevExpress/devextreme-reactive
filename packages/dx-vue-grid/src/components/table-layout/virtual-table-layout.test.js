import { shallow, mount } from '@vue/test-utils';
import { setupConsole } from '@devexpress/dx-testing';
import { getCollapsedGrid } from '@devexpress/dx-grid-core';
import { VirtualTableLayout } from './virtual-table-layout';

jest.mock('@devexpress/dx-grid-core', () => {
  const actual = require.requireActual('@devexpress/dx-grid-core');
  jest.spyOn(actual, 'getCollapsedGrid');
  return actual;
});
jest.mock('./column-group', () => ({
  ColumnGroup: () => null,
}));
jest.mock('@devexpress/dx-vue-core', () => ({
  DxRefHolder: {
    name: 'DxRefHolder',
    render() {
      return this.$slots.default[0];
    },
  },
  DxSizer: {
    name: 'DxSizer',
    render() {
      return this.$scopedSlots.default({ width: 800 });
    },
  },
}));

const defaultProps = {
  columns: [
    { key: 'a', column: { name: 'a' } },
    { key: 'b', column: { name: 'b' } },
    { key: 'c', column: { name: 'c' } },
    { key: 'd', column: { name: 'd' } },
    { key: 'e', column: { name: 'e' } },
  ],
  minWidth: 400,
  minColumnWidth: 120,
  height: 100,
  estimatedRowHeight: 40,
  bodyRows: [
    { key: 1 },
    { key: 2 },
    { key: 3 },
    { key: 4 },
    { key: 5 },
    { key: 6 },
  ],
  containerComponent: { render() { return <div>{this.$slots.default}</div>; } },
  headTableComponent: { render() { return <table>{this.$slots.default}</table>; } },
  tableComponent: { render() { return <table>{this.$slots.default}</table>; } },
  headComponent: { render() { return <table>{this.$slots.default}</table>; } },
  bodyComponent: { render() { return <tbody>{this.$slots.default}</tbody>; } },
  rowComponent: { name: 'Row', render() { return <tr>{this.$slots.default}</tr>; } },
  cellComponent: { name: 'Cell', render() { return <td />; } },
  getCellColSpan: () => 1,
};

describe('VirtualTableLayout', () => {
  let resetConsole;
  beforeEach(() => {
    resetConsole = setupConsole();
  });

  afterEach(() => {
    resetConsole();
    jest.clearAllMocks();
  });

  const simulateScroll = (tree, props) => {
    const target = {
      scrollTop: 0,
      scrollLeft: 0,
      ...props,
    };

    const eventData = {
      target,
      currentTarget: target,
    };

    // eslint-disable-next-line no-param-reassign
    tree.find(defaultProps.containerComponent).element.target = target;
    // eslint-disable-next-line no-param-reassign
    tree.find(defaultProps.containerComponent).element.currentTarget = target;

    tree.find(defaultProps.containerComponent).vm.$emit('scroll', eventData);
    tree.update();
  };

  it('should render correct layout', () => {
    const tree = shallow({
      render() {
        return (
          <VirtualTableLayout
            {...{ attrs: { ...defaultProps } }}
            headerRows={defaultProps.bodyRows.slice(0, 1)}
          />
        );
      },
    });

    expect(tree.vm.$el)
      .toMatchSnapshot();
  });

  describe('viewport', () => {
    it('should pass correct viewport at startup', () => {
      shallow({
        render() {
          return (
            <VirtualTableLayout
              {...{ attrs: { ...defaultProps } }}
              headerRows={defaultProps.bodyRows.slice(0, 1)}
            />
          );
        },
      });

      expect(getCollapsedGrid.mock.calls[getCollapsedGrid.mock.calls.length - 2][0])
        .toMatchObject({
          top: 0,
          left: 0,
          height: defaultProps.estimatedRowHeight,
          width: 800,
        });
      expect(getCollapsedGrid.mock.calls[getCollapsedGrid.mock.calls.length - 1][0])
        .toMatchObject({
          top: 0,
          left: 0,
          height: defaultProps.height - defaultProps.estimatedRowHeight,
          width: 800,
        });
    });

    it('should pass correct viewport at on viewport change', () => {
      const tree = mount({
        render() {
          return (
            <VirtualTableLayout
              {...{ attrs: { ...defaultProps } }}
              headerRows={defaultProps.bodyRows.slice(0, 1)}
            />
          );
        },
      });

      simulateScroll(tree, { scrollTop: 100, scrollLeft: 50 });

      expect(getCollapsedGrid.mock.calls[getCollapsedGrid.mock.calls.length - 2][0])
        .toMatchObject({
          top: 0,
          left: 50,
        });
      expect(getCollapsedGrid.mock.calls[getCollapsedGrid.mock.calls.length - 1][0])
        .toMatchObject({
          top: 100,
          left: 50,
        });
    });
  });

  describe('row heights', () => {
    it('should specify correct row height at startup', () => {
      expect.hasAssertions();

      const rows = [
        { key: 1 },
        { key: 2, height: 10 },
      ];

      getCollapsedGrid
        .mockImplementationOnce((args) => {
          const { getRowHeight } = args;
          expect(getRowHeight(rows[0]))
            .toEqual(defaultProps.estimatedRowHeight);
          expect(getRowHeight(rows[1]))
            .toEqual(10);

          return require.requireActual('@devexpress/dx-grid-core').getCollapsedGrid(args);
        });

      mount({
        render() {
          return (
            <VirtualTableLayout
              {...{ attrs: { ...defaultProps } }}
              bodyRows={rows}
            />
          );
        },
      });
    });

    it('should clear row height when rows updated', () => {
      const rows = [
        { key: 11 },
        { key: 12, height: 30 },
      ];

      const tree = mount(VirtualTableLayout, { propsData: { ...defaultProps, headerRows: rows } });
      tree.setProps({ bodyRows: [rows[0]] });

      const { getRowHeight } = getCollapsedGrid.mock.calls[0][0];
      expect(getRowHeight(rows[0]))
        .toEqual(defaultProps.estimatedRowHeight);
      expect(getRowHeight(rows[1]))
        .toEqual(30);
    });

    it('should clear row height when headerRows updated', () => {
      const rows = [
        { key: 11 },
        { key: 12 },
      ];

      const tree = mount(VirtualTableLayout, { propsData: { ...defaultProps, headerRows: rows } });
      tree.setProps({ headerRows: [rows[0]] });

      const { getRowHeight } = getCollapsedGrid.mock.calls[0][0];
      expect(getRowHeight(rows[1]))
        .toEqual(defaultProps.estimatedRowHeight);
    });
  });
});
