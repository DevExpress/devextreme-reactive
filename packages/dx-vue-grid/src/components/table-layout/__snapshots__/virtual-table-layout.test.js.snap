// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualTableLayout should render correct layout 1`] = `
<div
  style="height: 100px;"
>
  <table
    minwidth="400px"
    style="min-width: 400px;"
  >
    <!---->
    <table>
      <tr
        row="[object Object]"
        tablerow="[object Object]"
      >
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
      </tr>
    </table>
  </table>
  <table
    minwidth="400px"
    style="min-width: 400px;"
  >
    <!---->
    <tbody>
      <tr
        row="[object Object]"
        tablerow="[object Object]"
      >
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
      </tr>
      <tr
        row="[object Object]"
        tablerow="[object Object]"
      >
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
      </tr>
      <tr
        row="[object Object]"
        tablerow="[object Object]"
      >
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
      </tr>
      <tr
        row="[object Object]"
        tablerow="[object Object]"
      >
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
      </tr>
      <tr
        row="[object Object]"
        tablerow="[object Object]"
      >
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
        <td
          colspan="1"
          tablecolumn="[object Object]"
          tablerow="[object Object]"
        />
      </tr>
      <tr
        height="40px"
        row="[object Object]"
        tablerow="[object Object]"
      />
    </tbody>
  </table>
</div>
`;
